crates/ruff_python_parser/src/lexer.rs

```diff
@@ -1355,6 +1380,105 @@ mod tests {
             ]
         );
     }
+
+    #[test]
+    fn test_jupyter_magic_help_end() {
+        let source = r"
+?foo?
+??   foo?
+??   foo  ?
+?foo??
+??foo??
+???foo?
+???foo??
+??foo???
+???foo???
+?? \
+    foo?
+?? \
+?
+%foo?
+%foo??
+%%foo???"
+            .trim();
```

Comment:
For reference, the IPython transformer will transform the above code as the below where `pinfo` and `pinfo2` corresponds to `Help` and `Help2` magic kind:
```python
get_ipython().run_line_magic('pinfo', 'foo')
get_ipython().run_line_magic('pinfo', 'foo')
get_ipython().run_line_magic('pinfo2', '   foo  ?')
get_ipython().run_line_magic('pinfo2', 'foo')
get_ipython().run_line_magic('pinfo2', 'foo')
get_ipython().run_line_magic('pinfo', 'foo')
get_ipython().run_line_magic('pinfo2', 'foo')
get_ipython().run_line_magic('pinfo2', 'foo???')
get_ipython().run_line_magic('pinfo2', '?foo???')
get_ipython().run_line_magic('pinfo', 'foo')
get_ipython().run_line_magic('pinfo2', ' ?')
get_ipython().run_line_magic('pinfo', '%foo')
get_ipython().run_line_magic('pinfo2', '%foo')

# This is the only difference between IPython transformers and our lexer where
# our lexer will emit this as cell magic (`%%`) command
get_ipython().run_line_magic('%foo???', '')
```